lib/rules/no-unused-vars.js

```diff
@@ -102,19 +122,12 @@ module.exports = function(context) {
             }
         },
 
+        "JSXExpressionContainer": function(node) {
+            flagJSXAsUsed(node.expression.name);
```

Comment:
The code between `{}` is parsed as an expression. But you are right, those cases I mentioned are working.